### ISSUE TITLE
return values for u.uaccess.get_resources_with_explicit_access(X)

### ISSUE THREAD

**Issue by alvacouch** [hydroshare/hydroshare#889]
The code for the current get_resources_with_explicit_access is both used in the resource landing page and very confusing. It currently eliminates all immutable resources from all responses. This has to be wrong, but the question is "what is right?" 

There are two main questions: 
   1. Should group privilege be included? I think so. 
   2. Should the responses here be sensitive to the "immutable" flag, or not: 
       1. If not, then they will reflect just the OWNER, CHANGE, VIEW permissions without any bias. Users must use other routines to discover immutability. 
       2. If so, then how should permissions be changed for immutability: 
         * Should CHANGE -> VIEW? 
         * Should OWNER -> VIEW? 
       This -- it turns out -- is a difficult question from a semantic perspective. 

There are two concepts here: "recorded privilege" and "effective privilege". 
* If we report recorded privilege, we need an extra query to get effective privilege. 
* If we report effective privilege, we lose the recorded privilege and would need to get that back via another query as needed. 

It would seem that the path of least resistance would be to *always report recorded privilege*, because the modifiers "immutable", "published" are in the resource objects that are returned, so that further filtering is easy based upon them....

But that leaves me wondering why the code currently removes immutable resources? What issue does that address? How can we address it in another way?

**Comment by alvacouch** [comment]
From @dtarb via email:
As I understand it this function is used to identify resources to list on the MyResources page.  I do not think that whether the resource is immutable or not should matter.  However I may not understand all that this function does, so am stating my assumption for the context of my answer.  If my assumption is wrong, then the answer may change.

My assumption is that this function returns only a list of resources and that other attributes about these resources that are used in the My Resources page are obtained by other functions.  (The My Resources page seems to be using (at least)  the following attributes for each resource: Owned,  Change, View, Favorite,  Labels,  Type, Title, Owners,  Last Modified)

Given that this function returns just a list I think it should do the following
 1. When called with Owner.  Return a list of resources for which the user is the owner regardless of publication or mutability status.
 2.  When called with Change.  Return a list of resources for which the user has change or the user is a member of a group with change, regardless of publication or mutability status.  
 3.  When called with View.  Return a list of resources for which the user has View or the user is a member of a group with view, regardless of publication or mutability status.

By being returned on this list, it does not mean that the user can necessarily change the resource.  Because if a resource is immutable then the effective capability associated with Owner or Change is View.  I think that this means that I favor the "declared privilege" interpretation.

Now there are other questions that come up, that should be addressed separately if needed.  A case can be made that when a resource is formally published, that includes setting the immutable flag, that having change and view permissions becomes moot.  Thus the act of publishing could remove all change and view permissions.  If we did this then the permissions on the resource would have changed, so whether it is returned on this list would change accordingly and declared and effective privilege would be harmonized.  I am not in favor of this change, because it would then result in removing a resource from a MyResources page for a user that had change permissions, and was perhaps involved in working on the resource to get it ready for publication.  I prefer to leave the change and view permissions on published resources, essentially as a history/record.

This then opens yet another question.  Should it be possible for a published and immutable resource to change permissions.  This would amount to changing of declared privilege permissions, since the effective permissions are that the resource is accessible to all, but not changeable.  I think that the answer is yes.  So if there are multiple owners, owners can add or delete owners (not removing the last one).  Owners can set or remove change and view permissions.  For resources that are immutable the only consequences of these actions would be changing whether resources appear on MyResources page lists, but I prefer to leave this in the control of the users even after the resources is immutable.  This means that the setting and revoking of declared privilege parts of access control for a resource are not frozen, by a resource becoming immutable.

**Comment by dtarb** [comment]
See also #15

**Comment by alvacouch** [comment]
It turns out that this is much deeper than I thought. 
   1. I think that mixing group resources and user resources on the same page is going to get very, very confusing. One needs to associate them with a group, which is -- ostensibly -- just like another user.  
   2. I think that this functionality needs to be repeated for groups. There needs to be "get_resources_with_explicit_access" for groups as well. 

I think therefore that there are two options for resolving this: 
  1. Report real rather than effective privilege and then resolve it with a later filter. 
  2. Squash CHANGE and leave OWNER alone 

  real privilege   |   not immutable    | immutable
  ------------------  |  -----------------------  | ---------------
   OWNER         |  OWNER             | OWNER
   CHANGE       |  CHANGE            | VIEW
   VIEW             |  VIEW                  | VIEW

No matter what we do, we are going to need OWNER to be OWNER even if the resource is immutable. We'll have to determine whether OWNED resources are immutable or not via some other mechanism. 

I _think_ this is what was intended in the original code. It didn't do that.

**Comment by dtarb** [comment]
I do not share @alvacouch 's concern over confusion.  I am quite happy for the MyResources page to intermingle resources to which I have been given access explicitly as a user, and to which I have access by virtue of being member of a group that has been given access.  I would like tools to help me organize these resources, such as being able to put them in folders, or tag them.  There may be default folders that are useful, e.g. "Shared with me" and "Shared with my groups" which would be good default places for these resources to show up.  I would however like to also be able to put them in other places.  The paradigm that is attractive to me is that of Google Drive, where resources that I have been given access to show up first in Shared with me, but I can then add to my drive where I have the opportunity to organize into folders as I wish.

I prefer the term "Declared privilege" rather than real privilege, because real to me means what is actually in effect, so is the same as effective privilege.  Suppose a user has change privilege over a resource.  Then the resource is made immutable.  Their declared privilege on the resource is still "change", but their effective privilege on the resource is now "view".

**Comment by alvacouch** [comment]
But the google drive paradigm is what I am advocating. Resources show up on the group landing page because they are shared with the group; you can then favorite them to make them "yours".

**Comment by dtarb** [comment]
I am happy for you to implement as you think is best.  In Google things just show up in "Shared with me" and I am OK with that.

**Comment by alvacouch** [comment]
But the point is that it does not appear in "my documents" until you tell it to. It appears in a specially designated subfolder. 

I think there is a semantic conflation here that we need to get past. 
 * The access control system only determines what you can do with something. 
 * We are using it to determine where something appears in listings. 
 * Where things appear in listings is really the job of the labeling system (with folders, or just with labels).

My gut feeling is that just having all shared resources land in "Shared with me" is not going to cut it for our users. They are going to want to know *why* it was shared. The most obvious answers are
 1. because it is relevant to me outside of any group I'm in.  
 2. because it is relevant to some group I'm in.  

Right this second, resources appear in the main "my resources" at the moment of sharing. They do not appear in "shared with me" first.

**Comment by alvacouch** [comment]
So, I still have the major quandary of what immutable should do *right now*. I still don't know the intent of the UI concerning immutable. Is the following proposal reasonable?


  real privilege   |   not immutable    | immutable
  ------------------  |  -----------------------  | ---------------
   OWNER         |  OWNER             | OWNER
   CHANGE       |  CHANGE            | VIEW
   VIEW             |  VIEW                  | VIEW

**Comment by alvacouch** [comment]
(This routine was written specifically for @mauriel133 so I would expect that it is his final decision.)

**Comment by dtarb** [comment]
I am happy to wait for @mauriel133 to weigh in.

**Comment by mauriel133** [comment]
I don't remember writing this routine so I'm assuming it was written by @pkdash when we worked together with the My Resources page. Perhaps @pkdash can answer your questions.

**Comment by mauriel133** [comment]
I think the proposal above is reasonable. If a resource becomes immutable all edit permissions (except ownership) should appear as view.

**Comment by alvacouch** [comment]
The real issue here is a conflation between access control and the UI. 
* The access control system should never "lie". It should report what is going on. If someone ends up with VIEW, it should report that fact. 
* The UI, however, needs to categorize things in ways that are not so neatly determined. Someone may have VIEW, but also some remaining OWNER functions. And in the UI, it is important to note whether one has control (as an OWNER) of whether one has VIEW or not. 
* So, in the UI, it can be advantageous to say someone has OWNER when that doesn't mean as much as before. 
* But also, in the UI, it makes sense to tell someone that they no longer have CHANGE privilege, because the thing is immutable, so that now they have VIEW. 
* When someone inherits CHANGE from a group, and the resource becomes immutable, there is no question that they actually have VIEW now. There is no such thing as a group OWNER. 

This is the reasoning behind my proposal above. 

So, I think this is purely a UI issue. I'm willing to do whatever is easiest for @mauriel133 ...  for now I am implementing the above as a stopgap. Please let me know.

**Comment by horsburgh** [comment]
For the record:  @alvacouch 's proposal for how to handle immutable resources so they don’t disappear from the My Resouces page sounds reasonable.  And it’s a good switch that will have to be done automatically because our current publication workflow states that the “Edit” button for Published resources will be disabled, so users couldn’t change permissions on an published resource if they wanted to because they won’t be able to access the edit page where the access control features are.

**Comment by alvacouch** [comment]
For @mauriel133 : @dtarb is advocating that "my resources" should work "like Google Docs." In Google Docs, things that are shared with you don't appear in your "my documents" feed immediately. They appear in a separate folder called "Shared With Me". Then you can move them to your main documents feed by the equivalent of "claiming" them. At this point they are "unfiled documents" and you can later associate a folder with them. 

This functionality requires folders and is not a priority *yet*.  Many other things to do first.

**Comment by hyi** [comment]
@alvacouch @dtarb @horsburgh @mauriel133 @pkdash While working on resource publication, I have changed get_resources_with_explicit_access() to remove immutable filter so that it can return published resources with ownership retained for my resources page to display published resources accordingly. I also downgraded edit permissions to view permissions for all edit users to a resource after the resource is published and made immutable. I modified my resources page accordingly so that resources are displayed properly on my resources page after resources are published. Will deploy my resource publication work to a VM after I am done with resource publication work for viewing and testing and use it as a baseline for further comments.

**Comment by alvacouch** [comment]
@hyi @pkdash @dtarb @horsburgh @mauriel133  See issue #893. In the next version of access control, simply setting the immutable flag will do everything you need. This will be forthcoming very shortly.

**Comment by hyi** [comment]
@alvacouch Do you think you can get this done by end of next week? I have already changed access control API functions as needed in my resource publication branch, but most likely not system-wide. If you can get this done by end of next week, I will not spend time doing this duplicate work any more, but rather work on DOI assignment with crossref and other pieces for resource publication, then integrate my work on top of your system-wide work for issue #893 to handle permissions for immutable and published case. Let me know. Thanks. @dtarb @horsburgh @mauriel133 @pkdash

**Comment by alvacouch** [comment]
I can get the changes finished by the end of next week, but it will take a while for anyone to review the base PR  #867  upon which these changes depend. There are a lot of changes before this one. 
(and #867 is hard enough without making it more difficult to review).

**Comment by hyi** [comment]
@alvacouch understood. The changes I have made so far regarding permissions and access control are sufficient for my resource publication work, so my work is not contingent upon your access control work. We can get our respective work merged with develop whenever ready. We just need to make sure whichever work that will be merged later needs to take into account the other work that was already merged in to make sure the previous-merged code does not break. The changes I made in access control level are minimal to accommodate my resource publication work, so the later integration should not be too bad.
